### PR TITLE
Use GEMM for LinearRegressor and LinearClassifier operators to improve performance

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Use GEMM for LinearRegressor.
Update scores in-place with any post processor to avoid unnecessary allocations/copies.

**Motivation and Context**
Improve performance of SKL benchmarks.

**Validation**
Took 2 models generated from the skl2onnx unit tests that used LinearRegressor. One had 2 targets, the other had 1. Created inputs with batch sizes of 100, 500 and 2500. Ran each model for 30,000 iterations with onnxruntime_perf_test.exe using v1.1.2 vs. latest master with these changes.

Also tried to use the threadpool for the softmax calculations, but there wasn't a clear benefit to doing so. Would be better to switch to using the implementation in the Softmax operator instead if that turns out to be a performance issue.

Batch size | Version | SklearnLinearRegression-Dec4 | 2 targets 
-- | -- | -- | --
100 | v1.1.2 | Average time cost | 0.026868 
  | new | Average inference time cost | 0.00690207 
  |   | Speedup | 3.89 
  |   |   |   
500 | v1.1.2 | Average time cost | 0.131223 
  | new | Average inference time cost | 0.00898752 
  |   | Speedup | 14.60 
  |   |   |   
2500 | v1.1.2 | Average time cost | 0.592431
  | new | Average inference time cost | 0.0273765 
  |   | Speedup | 21.64 
  |   |   |   
Batch size | Version| SklearnLinearRegressionNoIntercept-Dec4 | 1 target |  
100 | v1.1.2 | Average time cost | 0.0152566 |  
  | new | Average inference time cost | 0.00501943 |  
  |   | Speedup | 3.04 |  
  |   |   |   |  
500 | v1.1.2 | Average time cost | 0.0583553 |  
  | new | Average inference time cost | 0.00738535 |  
  |   | Speedup | 7.90 |  
  |   |   |   |  
2500 | v1.1.2 | Average time cost | 0.275574 |  
  | new | Average inference time cost | 0.018674 |  
  |   | Speedup | 14.76 |

### CODE REVIEW

File: onnxruntime/core/providers/cpu/ml/ml_common.h [hunk 1]
```diff
@@ -342,5 +356,72 @@ void write_scores(std::vector<T>& scores, POST_EVAL_TRANSFORM post_transform, in
...
+
+  ORT_UNUSED_PARAMETER(add_second_class);  // pending updates to use batched_update_scores_inplace for more ops
+  ORT_UNUSED_PARAMETER(threadpool);        // TBD whether we need to parallelize code here
+
+  // convert from span to pointer for efficiency. we've checked scores.size() matches num_scores so don't need the
+  // extra checking/overhead from using operator[] for each access
+  T* s = &*scores.begin();
+  const T* s_end = &*scores.cend();
+
+  if (batch_size > 1) {
```
Comment (prabhat00155) [inline onnxruntime/core/providers/cpu/ml/ml_common.h]: You can remove else if part by changing the if condition to:
```
if (batch_size > 1 || post_transform == POST_EVAL_TRANSFORM::PROBIT)
```
I assume batch size >= 1.

Comment (skottmckay) [reply onnxruntime/core/providers/cpu/ml/ml_common.h]: I have the changes to convert LinearClassifier to use GEMM as well, and those add all the handling for 'add_second_class' in the 'else' (see line 332 - 345 above), so any removal here would be temporary. 

---
In reply to: [389724122](https://github.com/microsoft/onnxruntime/pull/3154#discussion_r389724122) [](ancestors = 389724122)

**Comment by skottmckay** [comment]
Added LinearClassifier changes as well. Tested speedup using SklearnRidgeClassifierMulti. If there are other skl2onnx test cases that should be checked please let me know.


Batch size | Speedup
-- | --
100 | 5.33
500 | 6.94
1500 | 11.56
2500 | 13.55
5000 | 17.95
